### PR TITLE
only hard deps in pre-commit - roxygenize - additional deps

### PR DESCRIPTION
**Pull request by pawelru**

### CODE REVIEW

**Comment by m7pr** [comment]
Hey @pawelru happy to review the issue. Can you tell what those pre-commit hooks do, and what your change introduces?

**Comment by pawelru** [comment]
> Hey @pawelru happy to review the issue. Can you tell what those pre-commit hooks do, and what your change introduces?

You have the description here: https://lorenzwalthert.github.io/precommit/articles/available-hooks.html

My changes only relates to one of the hooks that needs to create temp env on which it will run the roxygen docs generation. When I ran this locally it failed on some non-relevant package unavailability as shown below. After the changes that's not the case anymore

```
   ~/Doc/r/gh/in/teal    readme_updates *3 +1 ········· 45s   base  
❯ pre-commit clean
Cleaned /Users/ruckip/.cache/pre-commit.

    ~/Doc/r/gh/in/teal    readme_updates *3 +1 ················· base  
❯ pre-commit run  
[INFO] Initializing environment for https://github.com/lorenzwalthert/precommit.
[INFO] Initializing environment for https://github.com/lorenzwalthert/precommit:checkmate,digest,dplyr,bioc::GenomeInfoDbData,ggplot2,glue,grDevices,lifecycle,logger,magrittr,methods,bioc::MultiAssayExperiment,R6,rlang,rtables,shinyjs,shinyWidgets,stats,bioc::SummarizedExperiment,utils,yaml.
[INFO] Installing environment for https://github.com/lorenzwalthert/precommit.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/lorenzwalthert/precommit.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
An unexpected error has occurred: CalledProcessError: command: ('/usr/local/bin/Rscript', '--no-save', '--no-restore', '--no-site-file', '--no-environ', '/var/folders/m1/hrz0h_ls7gz57rc80tnj41t80000gp/T/tmpx061b2b2/script.R', 'checkmate', 'digest', 'dplyr', 'bioc::GenomeInfoDbData', 'ggplot2', 'glue', 'grDevices', 'lifecycle', 'logger', 'magrittr', 'methods', 'bioc::MultiAssayExperiment', 'R6', 'rlang', 'rtables', 'shinyjs', 'shinyWidgets', 'stats', 'bioc::SummarizedExperiment', 'utils', 'yaml')
return code: 1
stdout:
    Retrieving 'https://packagemanager.rstudio.com/all/latest/bin/macosx/big-sur-x86_64/contrib/4.3/checkmate_2.2.0.tgz' ...
        OK [file is up to date]
    Retrieving 'https://packagemanager.rstudio.com/all/latest/bin/macosx/big-sur-x86_64/contrib/4.3/dplyr_1.1.2.tgz' ...
        OK [file is up to date]
    Retrieving 'https://packagemanager.rstudio.com/all/latest/bin/macosx/big-sur-x86_64/contrib/4.3/generics_0.1.3.tgz' ...
        OK [file is up to date]
    Retrieving 'https://packagemanager.rstudio.com/all/latest/bin/macosx/big-sur-x86_64/contrib/4.3/pillar_1.9.0.tgz' ...
        OK [file is up to date]
    Retrieving 'https://packagemanager.rstudio.com/all/latest/bin/macosx/big-sur-x86_64/contrib/4.3/fansi_1.0.4.tgz' ...
        OK [file is up to date]
    Retrieving 'https://packagemanager.rstudio.com/all/latest/bin/macosx/big-sur-x86_64/contrib/4.3/utf8_1.2.3.tgz' ...
        OK [file is up to date]
    Retrieving 'https://packagemanager.rstudio.com/all/latest/bin/macosx/big-sur-x86_64/contrib/4.3/tibble_3.2.1.tgz' ...
        OK [file is up to date]
    Retrieving 'https://packagemanager.rstudio.com/all/latest/bin/macosx/big-sur-x86_64/contrib/4.3/pkgconfig_2.0.3.tgz' ...
        OK [file is up to date]
    Retrieving 'https://packagemanager.rstudio.com/all/latest/bin/macosx/big-sur-x86_64/contrib/4.3/tidyselect_1.2.0.tgz' ...
        OK [file is up to date]
    Installing BiocManager [1.30.21.1] ...
        OK [linked cache in 0.61 milliseconds]
    * Installed 1 package in 21 milliseconds.
stderr:
    Error: package 'GenomeInfoDbData' is not available
    In addition: Warning message:
    no MRAN records available from repository URL '/bin/macosx/big-sur-x86_64/contrib/4.3' 
    Traceback (most recent calls last):
    7: renv::install(commandArgs(trailingOnly = TRUE))
    6: retrieve(names(remotes))
    5: handler(package, renv_retrieve_impl(package))
    4: renv_retrieve_impl(package)
    3: renv_available_packages_latest(package)
    2: stopf("package '%s' is not available", package)
    1: stop(sprintf(fmt, ...), call. = call.)
    Execution halted
Check the log at /Users/ruckip/.cache/pre-commit/pre-commit.log
```

**Comment by m7pr** [comment]
will try to install python, to get pip to install pre-commit to review this locally. is there any docker container that we might use for testing this?

**Comment by m7pr** [comment]
Currently having issues installing Python + miniconda on my Windows machine. Will update you when I get this sorted out, or will try Docker with Python to review this.

**Comment by donyunardi** [comment]
Once you have pre-commit locally, you'd run `pre-commit install` on the repo with pre-commit yaml file.
This will automate the pre-commit run upon commit.

I ran it locally and got an error about teal.data, teal.slice, and teal.transform.

```unix
❯ pre-commit clean
Cleaned /Users/unardid/.cache/pre-commit.
❯ git add .
❯ git commit -m "test"
[INFO] Initializing environment for https://github.com/lorenzwalthert/precommit.
[INFO] Initializing environment for https://github.com/lorenzwalthert/precommit:checkmate,lifecycle,logger,magrittr,rlang,shinyjs,stats,insightsengineering/teal.logger,insightsengineering/teal.reporter,insightsengineering/teal.widgets,utils.
[INFO] Installing environment for https://github.com/lorenzwalthert/precommit.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/lorenzwalthert/precommit.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
Style code with `styler`.................................................Failed
- hook id: style-files
- files were modified by this hook

ℹ Using R 4.2.2 (lockfile was generated with R 4.3.0)
ℹ Using R 4.2.2 (lockfile was generated with R 4.3.0)
NULL
[1] "cache root set to " "styler"            
Styling  1  files:
 R/utils.R ℹ 
────────────────────────────────────────
Status  Count   Legend 
✔       0       File unchanged.
ℹ       1       File changed.
✖       0       Styling threw an error.
────────────────────────────────────────
Please review the changes carefully!


Regenerate package documentation.........................................Failed
- hook id: roxygenize
- exit code: 1

ℹ Using R 4.2.2 (lockfile was generated with R 4.3.0)
ℹ Using R 4.2.2 (lockfile was generated with R 4.3.0)
ℹ Loading teal
Error in `precommit::roxygenize_with_cache()`:
! The packages "teal.data" (>= 0.3.0), "teal.slice" (>= 0.4.0), and
"teal.transform" (>= 0.4.0) are required.
Backtrace:
    ▆
 1. └─precommit::roxygenize_with_cache(key = wd, dirs = path_relative_cache)
 2.   └─rlang::abort(conditionMessage(out))
Execution halted


Check spelling with `spelling`.......................(no files to check)Skipped
```

These three packages are in `Depends` so I figure we probably need to add these into the dependencies list.

However, I ran into the usual `API rate limit exceeded` from Github when I run the pre-commit again, stopping me to install all teal dependencies.

```unix
    Error: failed to resolve remote 'insightsengineering/teal.logger' -- error downloading 'https://api.github.com/repos/insightsengineering/teal.logger' [curl: (22) The requested URL returned error: 403]
    Traceback (most recent calls last):
    18: renv::install(commandArgs(trailingOnly = TRUE))
    17: lapply(remotes, renv_remotes_resolve)
    16: FUN(X[[i]], ...)
    15: withCallingHandlers(renv_remotes_resolve_impl(spec, latest), 
            error = error)
    14: renv_remotes_resolve_impl(spec, latest)
    13: renv_remotes_resolve_github(remote)
    12: remote$ref %||% renv_remotes_resolve_github_ref(host, user, repo)
    11: renv_remotes_resolve_github_ref(host, user, repo)
    10: tryCatch(renv_remotes_resolve_github_ref_impl(host, user, repo), 
            error = function(e) {
                warning(e)
                getOption("renv.github.default_branch", default = "master")
            })
     9: tryCatchList(expr, classes, parentenv, handlers)
     8: tryCatchOne(expr, names, parentenv, handlers[[1L]])
     7: value[[3L]](cond)
     6: warning(e)
     5: withRestarts({
            .Internal(.signalCondition(cond, message, call))
            .Internal(.dfltWarn(message, call))
        }, muffleWarning = function() NULL)
     4: withOneRestart(expr, restarts[[1L]])
     3: doWithOneRestart(return(expr), restart)
     2: <condition-handler>(...)
     1: stop(simpleError(message = message, call = e$call))
    Execution halted
Check the log at /Users/unardid/.cache/pre-commit/pre-commit.log
```

Not sure how to pass authentication from pre-commit.
Maybe we should try to install from r-universe?

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -16,25 +16,16 @@ repos:
...
-          - R6
           - rlang
-          - rtables
           - shinyjs
-          - shinyWidgets
           - stats
-          - bioc::SummarizedExperiment
+          - insightsengineering/teal.logger
+          - insightsengineering/teal.reporter
+          - insightsengineering/teal.widgets
```
Comment (donyunardi) [inline .pre-commit-config.yaml:27]: also add:
- insightsengineering/teal.data
- insightsengineering/teal.slice
- insightsengineering/teal.transform

**Comment by donyunardi** [comment]
> Not sure how to pass authentication from pre-commit.

I attempted to execute the process again after waiting for a few hours, and this time it succeeded. I assume that the rate limit is reset after a certain period of time.

Regarding the output provided below, I was in the process of adding a new function to the utils.R with bad formatting and didn't generate the corresponding .Rd file for the function.

```
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   R/utils.R

❯ git add .
❯ git commit -m "test again"
Style code with `styler`.................................................Failed
- hook id: style-files
- files were modified by this hook

ℹ Using R 4.2.2 (lockfile was generated with R 4.3.0)
ℹ Using R 4.2.2 (lockfile was generated with R 4.3.0)
NULL
[1] "cache root set to " "styler"            
Styling  1  files:
 R/utils.R ℹ 
────────────────────────────────────────
Status  Count   Legend 
✔       0       File unchanged.
ℹ       1       File changed.
✖       0       Styling threw an error.
────────────────────────────────────────
Please review the changes carefully!


Regenerate package documentation.........................................Failed
- hook id: roxygenize
- files were modified by this hook

ℹ Using R 4.2.2 (lockfile was generated with R 4.3.0)
ℹ Using R 4.2.2 (lockfile was generated with R 4.3.0)

You are using teal version 0.14.0.9000
ℹ Loading teal

You are using teal version 0.14.0.9000
Writing NAMESPACE


Check spelling with `spelling`.......................(no files to check)Skipped
```

File: .pre-commit-config.yaml (hunk 1 above)
Comment (pawelru) [reply .pre-commit-config.yaml:27]: good catch - thanks!

**Comment by pawelru** [comment]
Thanks Dony. There are couple of things in what you showed:

```
ℹ Using R 4.2.2 (lockfile was generated with R 4.3.0)
```
Please use R 4.3.1 as in our CI golden image. Theoretically the results might differ because of that but I personally doubt it is true in our case.

```
API rate limit exceeded
```
Yes you are right. Installing packages from GH is limited **unless you specify GH PAT**. pre-commit is calling a new R process and it looks like this is not accessible from within that new process. Please set up global system env or modify your R profile. 
From what I recall this limit get reset every hour.

**Review by averissimo — COMMENTED** [review]
Q: How about adding the `lintr` hook for the changed files?

Good to see this PR :+1: , are these changes being applied to other repositories?

I was having the same issues with `pre-commit` :) described here: https://github.com/lorenzwalthert/precommit/issues/507

**Comment by pawelru** [comment]
> Q: How about adding the lintr hook?

Good idea but I don't want to explode this PR. I would prefer to do this within https://github.com/insightsengineering/coredev-tasks/issues/77

> are these changes being applied to other repositories?

Nope. For now just `teal` repo but once we have an agreement I will create a separate issue to apply similar changes elsewhere and treat this PR as a reference.

**Review by donyunardi — APPROVED** [review]
LGTM.
Probably need to add a section in onboarding docs on how to install pre-commit locally and in the repo.

**Pull request merged**